### PR TITLE
Initial Opinfo Testing Framework

### PR DESCRIPTION
**Pull request by rdspring1**
This PR add support for an `opinfo` testing framework using `pytest`.
Run tests with `pytest python_tests/pytest_ops.py`

Files:
1. pytest_framework.py
4. pytest_core.py
5. pytest_input_generators.py
6. pytest_opinfos.py
7. pytest_ops.py
8. pytest_utils.py

Test Templates:
1. `test_errors` - Runs operations with error inputs and expects an Exception with desired message.
2. `test_correctness` - Compares NvFuser results against a reference function.
3. `test_definition_op_in_schedule_error` - Checks that fusion operations are not called in `schedule` function.

Tested Operations:
1. acos
2. slice
3. define_tensor
4. var_mean

### CODE REVIEW

**Review by jjsjann123 — COMMENTED** [review]
Notice that we are lacking license header on added files. Wasn't sure if that's intended, since there's the make_tensor thing coming from pytorch.

A quick question is, what's the jax reference about here? Sorry if I'm jus totally not aware of the context here.

File: python_tests/pytest_core.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+from functools import partial, wraps
+from typing import Callable
+import torch
+import itertools
+from enum import Enum
+
+
+class ReferenceType(Enum):
+    Pytorch = 1
+    Jax = 2
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply python_tests/pytest_core.py:12]: It is sometimes easier to test our functions against Jax. Thunder has this option too.
- [shift_right_logical](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.shift_right_logical.html)
- [slice](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html)

**Comment by rdspring1** [comment]
There are some prim functions that correspond to Jax functions.
- [shift_right_logical](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.shift_right_logical.html)
- [slice](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html)

File: python_tests/pytest_core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjsjann123) [reply python_tests/pytest_core.py:12]: so we'll want `jax` in our `requirements.txt`. Maybe we should have it split up in different files/sections.

File: csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -2282,7 +2282,7 @@ void initNvFuserPythonBindings(PyObject* module) {
       },
       py::arg("arg"),
       py::arg("axes"),
-      py::arg("correction"),
+      py::arg("correction") = 1,
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:2285]: Added default parameter to match pytorch signature.
https://pytorch.org/docs/stable/generated/torch.var_mean.html

**Comment by rdspring1** [comment]
@kevinstephano @jjsjann123 This PR is ready for review.

File: python_tests/pytest_core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinstephano) [reply python_tests/pytest_core.py]: It's okay to use jax to verify operations.

File: python_tests/pytest_input_generators.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+    return _fn
+
+
+def define_tensor_generator(op, dtype, requires_grad, **kwargs):
+    yield SampleInput(symbolic_sizes=[-1], contiguity=[True])
+
+
+def define_tensor_error_generator(op, dtype, requires_grad, **kwargs):
+    """
```
Comment (kevinstephano) [inline python_tests/pytest_input_generators.py:88]: I would add a few more checks:
1. `symbolic_sizes` should check that non-zeros trigger errors
2.  `contiguity` should be checked for values that are not `None`, `True`, or `False`.

Comment (kevinstephano) [reply python_tests/pytest_input_generators.py:88]: Or at least add a TODO.  This isn't a blocker for getting in the basic framework.

**Review by kevinstephano — COMMENTED** [review]
Looks good to me except (1.) comment needs to be removed and a TODO could be added for `define_tensor` checking.  I would like @jjsjann123 to provide the approval stamp just to confirm everything is in order for the build and CI.

File: python_tests/pytest_input_generators.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    # Typical inputs
+    for shape in shapes:
+        yield SampleInput(make_arg(shape))
+
+    # Noncontiguous inputs
+    for shape in shapes:
+        yield SampleInput(make_arg(shape, noncontiguous=True))
+
+    # Arbitrarily strided inputs
```
Comment (jjsjann123) [inline python_tests/pytest_input_generators.py:55]: Just a note, is overlapping memory something we want to support?

**Comment by jjsjann123** [comment]
!build

File: python_tests/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Installs the wheel compatible with CUDA 12 and cuDNN 8.9 or newer.
+# Note: wheels only available on linux.
+jax[cuda12_local]
```
Comment (jjsjann123) [inline python_tests/requirements.txt:3]: note that we do build against cuda11.7/cuda11.8 as well. So might not want to have a hard-coded cuda12 here.

if jax isn't needed for build, we can have the requirement listed in `setup.py`

**Review by jjsjann123 — COMMENTED** [review]
I think any pytest file under `python_tests` would just be ignored.. But let's see how github CI looks like with these changes.

cc'ing on @xwang233 , in case this accidentally affected nightly CI.

**Comment by xwang233** [comment]
We currently have `python_tests/test_*.py` python tests running in CI, so those pytest tests won't be running for now. Although we can add more rules for python tests, it's still recommended to add a universal test entry point, similar to those in pytorch, e.g. `python run_test.py`.

cc @jjsjann123

**Comment by rdspring1** [comment]
!build

File: python_tests/pytest_input_generators.py (hunk 4 above)
Comment (rdspring1) [reply python_tests/pytest_input_generators.py:55]: Probably could remove these samples.

File: python_tests/requirements.txt (hunk 5 above)
Comment (jjsjann123) [reply python_tests/requirements.txt:3]: Yeah, but it should match the CUDA TK version on the system. ie. `jax[cuda11_local]` for cuda 11.7/8.

Comment (rdspring1) [reply python_tests/requirements.txt:3]: Can you do this automatically in the pip `requirements.txt`? file

Comment (jjsjann123) [reply python_tests/requirements.txt:3]: Oops, sorry that I missed this comment.
I don't know if there's any way to do that in requirements.txt. But we are also not really using jax for build (unlike pybind & ninja), so it doesn't need to be installed with the requirements.txt (I think?!)

For pip wheel, we can have this as one of our `install_requires` (setup.py), which we will update in our pip building script.
For developers, well, we can just leave it out and ask people to install jax as a dependency. (maybe put this in the build wiki on github).

All I'm saying here is, let's just delete this file :laughing:

**Comment by jjsjann123** [comment]
FYI, main is broken on nvfuser_bench.

**Comment by naoyam** [comment]
I'm hearing multiple reports of benchmark failures. Could somebody please file an issue with a repro?

**Comment by jjsjann123** [comment]
> I'm hearing multiple reports of benchmark failures. Could somebody please file an issue with a repro?

https://github.com/NVIDIA/Fuser/issues/472

**Comment by naoyam** [comment]
Thanks! Didn't you also mentioned a timeout happened?

**Comment by naoyam** [comment]
Never mind, saw your response to the other PR

File: python_tests/pytest_input_generators.py (hunk 4 above)
Comment (rdspring1) [reply python_tests/pytest_input_generators.py]: I removed them.

File: python_tests/requirements.txt (hunk 5 above)
Comment (rdspring1) [reply python_tests/requirements.txt]: I removed the `requirements.txt` file and added a `Pytest.md` file.

**Review by jjsjann123 — APPROVED** [review]
I'm stamping this one as-is. since it isn't really used in CI yet and won't affect anything regarding build.

Thanks for taking the first step on refactoring functional/error-handling testing.

**Pull request merged**